astropy/io/fits/hdu/image.py

```diff
@@ -827,6 +827,11 @@ def _get_scaled_image_data(self, offset, shape):
         data = None
         if not (self._orig_bzero == 0 and self._orig_bscale == 1):
             data = self._convert_pseudo_integer(raw_data)
+            if not (self._blank is None or data is None or data.dtype.kind != "u"):
+                raw_data = data
+                data = None
+                self._uint = False  # Reset this to enable conversion to float
+                self._blank += self._orig_bzero  # Same scaling to uint as for data
```

Comment:
Hi @dhomeier,

Sorry it took me too long to come to this PR but I was focused on the compression refactor :)
I would prefer to avoid the few lines above which are making this function harder to understand (tweaking the hdu to make it behaves as integer data). I think we can handle directly the blank case for unsigned integer in `_convert_pseudo_integer`, something like this:

```
--- astropy/io/fits/hdu/image.py
+++ astropy/io/fits/hdu/image.py
@@ -795,6 +795,11 @@ class _ImageBaseHDU(_ValidHDU):
             data = np.array(data, dtype=dtype)
             data -= np.uint64(1 << (bits - 1))
 
+            if self._blank is not None:
+                data = np.array(data, dtype=np.float32)
+                blanks = data.flat == self._blank + self._orig_bzero
+                data.flat[blanks] = np.nan
+
             return data
 
     def _get_scaled_image_data(self, offset, shape):
@@ -827,11 +832,6 @@ class _ImageBaseHDU(_ValidHDU):
         data = None
         if not (self._orig_bzero == 0 and self._orig_bscale == 1):
             data = self._convert_pseudo_integer(raw_data)
-            if not (self._blank is None or data is None or data.dtype.kind != "u"):
-                raw_data = data
-                data = None
-                self._uint = False  # Reset this to enable conversion to float
-                self._blank += self._orig_bzero  # Same scaling to uint as for data
```

Does is it reasonable ?